mypy/checkexpr.py

```diff
@@ -1375,8 +1376,67 @@ def visit_type_application(self, tapp: TypeApplication) -> Type:
 
     def visit_type_alias_expr(self, alias: TypeAliasExpr) -> Type:
+        """ Get type of a type alias (could be generic) in a runtime expression."""
+        item = alias.type
+        if (isinstance(item, (Instance, TupleType, UnionType, CallableType))
+           and not alias.runtime):
+            item = self.replace_tvars_any(item)
+        if isinstance(item, Instance):
+            tp = type_object_type(item.type, self.named_type)
+        else:
+            # TODO: Better error reporting: need to find line for
+            # unsubscribed generic aliases, that are invalid at runtime.
+            if alias.line > 0:
+                self.chk.fail('Invalid type alias in runtime expression: {}'
+                              .format(item), alias)
+            return AnyType()
+        if isinstance(tp, CallableType):
+            if len(tp.variables) != len(item.args):
+                self.msg.incompatible_type_application(len(tp.variables),
+                                                       len(item.args), item)
+                return AnyType()
+            return self.apply_generic_arguments(tp, item.args, item)
+        elif isinstance(tp, Overloaded):
+            for it in tp.items():
+                if len(it.variables) != len(item.args):
+                    self.msg.incompatible_type_application(len(it.variables),
+                                                           len(item.args), item)
+                    return AnyType()
+            return Overloaded([self.apply_generic_arguments(it, item.args, item)
+                               for it in tp.items()])
         return AnyType()
 
+    def replace_tvars_any(self, tp: Type) -> Type:
+        """ Replace all unbound type variables with Any if an alias is used in
+        a runtime expression. Basically, this function finishes what could not be done
+        in similar funtion from typeanal.py.
+        """
+        if not isinstance(tp, (Instance, UnionType, TupleType, CallableType)):
+            return tp
+        typ_args = (tp.args if isinstance(tp, Instance) else
+                    tp.items if not isinstance(tp, CallableType) else
+                    tp.arg_types + [tp.ret_type])
+        new_args = typ_args[:]
+        for i, arg in enumerate(typ_args):
+            if isinstance(arg, UnboundType):
+                sym = None
+                try:
+                    sym = self.chk.lookup_qualified(arg.name)
+                except KeyError:
+                    pass
+                if sym and (sym.kind == UNBOUND_TVAR or sym.kind == BOUND_TVAR):
+                    new_args[i] = AnyType()
+            else:
+                new_args[i] = self.replace_tvars_any(arg)
+        if isinstance(tp, Instance):
+            return Instance(tp.type, new_args, tp.line)
+        if isinstance(tp, TupleType):
+            return tp.copy_modified(items=new_args)
+        if isinstance(tp, UnionType):
+            return UnionType.make_union(new_args, tp.line)
```

Comment:
@gvanrossum OK, I did this. Note that because of this change ``Union[Any, Node[T]][int]`` is simplified to ``Any``.